In [1]:
# Supress Warnings

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

## Task 1: Reading and Inspection

-  ### Subtask 1.1: Import and read

Import and read the movie database. Store it in a variable called `movies`.

In [3]:
movies = pd.read_csv('Movie Dataset.csv')
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
5038  ...                  6.0  English   Canada             NaN          NaN   
5039  ...                359.0  English      USA           TV-14          NaN   
5040  ...                  3.0  English      USA             NaN       1400.0   
5041  ...                  9.0  English      USA           PG-13          NaN   
5042  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

-  ### Subtask 1.2: Inspect the dataframe

Inspect the dataframe's columns, shapes, variable types etc.

In [4]:
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [5]:
movies.shape

(5043, 28)

In [6]:
movies.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             4993.000000  5028.000000              4939.000000   
mean               140.194272   107.201074               686.509212   
std                121.601675    25.197441              2813.328607   
min                  1.000000     7.000000                 0.000000   
25%                 50.000000    93.000000                 7.000000   
50%                110.000000   103.000000                49.000000   
75%                195.000000   118.000000               194.500000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             5020.000000             5036.000000  4.159000e+03   
mean               645.009761             6560.047061  4.846841e+07   
std               1665.041728            15020.759120  6.845299e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                133.000000              614.000000  5.340988e+06   
50%                371.500000              988.000000  2.551750e+07   
75%                636.000000            11000.000000  6.230944e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     5.043000e+03                5043.000000           5030.000000   
mean      8.366816e+04                9699.063851              1.371173   
std       1.384853e+05               18163.799124              2.013576   
min       5.000000e+00                   0.000000              0.000000   
25%       8.593500e+03                1411.000000              0.000000   
50%       3.435900e+04                3090.000000              1.000000   
75%       9.630900e+04               13756.500000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           5022.000000  4.551000e+03  4935.000000   
mean             272.770808  3.975262e+07  2002.470517   
std              377.982886  2.061149e+08    12.474599   
min                1.000000  2.180000e+02  1916.000000   
25%               65.000000  6.000000e+06  1999.000000   
50%              156.000000  2.000000e+07  2005.000000   
75%              326.000000  4.500000e+07  2011.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             5030.000000  5043.000000   4714.000000           5043.000000  
mean              1651.754473     6.442138      2.220403           7525.964505  
std               4042.438863     1.125116      1.385113          19320.445110  
min                  0.000000     1.600000      1.180000              0.000000  
25%                281.000000     5.800000      1.850000              0.000000  
50%                595.000000     6.600000      2.350000            166.000000  
75%                918.000000     7.200000      2.350000           3000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

In [7]:
movies.count()

color                        5024
director_name                4939
num_critic_for_reviews       4993
duration                     5028
director_facebook_likes      4939
actor_3_facebook_likes       5020
actor_2_name                 5030
actor_1_facebook_likes       5036
gross                        4159
genres                       5043
actor_1_name                 5036
movie_title                  5043
num_voted_users              5043
cast_total_facebook_likes    5043
actor_3_name                 5020
facenumber_in_poster         5030
plot_keywords                4890
movie_imdb_link              5043
num_user_for_reviews         5022
language                     5029
country                      5038
content_rating               4740
budget                       4551
title_year                   4935
actor_2_facebook_likes       5030
imdb_score                   5043
aspect_ratio                 4714
movie_facebook_likes         5043
dtype: int64

## Task 2: Cleaning the Data

-  ### Subtask 2.1: Inspect Null values

Find out the number of Null values in all the columns and rows. Also, find the percentage of Null values in each column. Round off the percentages upto two decimal places.

In [8]:
# Write your code for column-wise null count here
movies.isnull().sum(axis= 0)

color                         19
director_name                104
num_critic_for_reviews        50
duration                      15
director_facebook_likes      104
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross                        884
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                153
movie_imdb_link                0
num_user_for_reviews          21
language                      14
country                        5
content_rating               303
budget                       492
title_year                   108
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 329
movie_facebook_likes           0
dtype: int64

In [9]:
# Write your code for row-wise null count here
movies.isnull().sum(axis=1)

0        0
1        0
2        0
3        0
4       14
        ..
5038     4
5039     5
5040     4
5041     2
5042     0
Length: 5043, dtype: int64

In [10]:
# Write your code for column-wise null percentages here
round(100*(movies.isnull().sum(axis=0)/len(movies)),2)

color                         0.38
director_name                 2.06
num_critic_for_reviews        0.99
duration                      0.30
director_facebook_likes       2.06
actor_3_facebook_likes        0.46
actor_2_name                  0.26
actor_1_facebook_likes        0.14
gross                        17.53
genres                        0.00
actor_1_name                  0.14
movie_title                   0.00
num_voted_users               0.00
cast_total_facebook_likes     0.00
actor_3_name                  0.46
facenumber_in_poster          0.26
plot_keywords                 3.03
movie_imdb_link               0.00
num_user_for_reviews          0.42
language                      0.28
country                       0.10
content_rating                6.01
budget                        9.76
title_year                    2.14
actor_2_facebook_likes        0.26
imdb_score                    0.00
aspect_ratio                  6.52
movie_facebook_likes          0.00
dtype: float64

-  ### Subtask 2.2: Drop unecessary columns

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [11]:
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations 
movies.drop(['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes',
             'actor_3_name', 'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link', 'aspect_ratio', 'plot_keywords'], axis = 1, inplace=True)
movies.head()

director_name  num_critic_for_reviews        gross  \
0      James Cameron                   723.0  760505847.0   
1     Gore Verbinski                   302.0  309404152.0   
2         Sam Mendes                   602.0  200074175.0   
3  Christopher Nolan                   813.0  448130642.0   
4        Doug Walker                     NaN          NaN   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4                      Documentary      Doug Walker   

                                         movie_title  num_voted_users  \
0                                            Avatar            886204   
1          Pirates of the Caribbean: At World's End            471220   
2                                           Spectre            275868   
3                             The Dark Knight Rises           1144337   
4  Star Wars: Episode VII - The Force Awakens    ...                8   

   num_user_for_reviews language       budget  title_year  imdb_score  \
0                3054.0  English  237000000.0      2009.0         7.9   
1                1238.0  English  300000000.0      2007.0         7.1   
2                 994.0  English  245000000.0      2015.0         6.8   
3                2701.0  English  250000000.0      2012.0         8.5   
4                   NaN      NaN          NaN         NaN         7.1   

   movie_facebook_likes  
0                 33000  
1                     0  
2                 85000  
3                164000  
4                     0

In [12]:
round(100*(movies.isnull().sum(axis=0)/len(movies)),2)

director_name              2.06
num_critic_for_reviews     0.99
gross                     17.53
genres                     0.00
actor_1_name               0.14
movie_title                0.00
num_voted_users            0.00
num_user_for_reviews       0.42
language                   0.28
budget                     9.76
title_year                 2.14
imdb_score                 0.00
movie_facebook_likes       0.00
dtype: float64

-  ### Subtask 2.3: Drop unecessary rows using columns with high Null percentages

Now, on inspection you might notice that some columns have large percentage (greater than 5%) of Null values. Drop all the rows which have Null values for such columns.

In [13]:
# Method 1
# movies = movies[~np.isnan(movies['gross'])]
# movies = movies[~np.isnan(movies['budget'])]

# Method 2
high_null_cols = movies.columns[movies.isnull().mean() * 100 > 5]
movies = movies.dropna(subset=high_null_cols)

In [14]:
round(100*(movies.isnull().sum(axis=0)/len(movies)),2)

director_name             0.00
num_critic_for_reviews    0.03
gross                     0.00
genres                    0.00
actor_1_name              0.08
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
language                  0.10
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64

In [15]:
movies.reindex()
movies.shape

(3891, 13)

-  ### Subtask 2.4: Drop unecessary rows

Some of the rows might have greater than five NaN values. Such rows aren't of much use for the analysis and hence, should be removed.

In [16]:
movies.isnull().sum(axis = 1)

0       0
1       0
2       0
3       0
5       0
       ..
5033    0
5034    0
5035    0
5037    0
5042    0
Length: 3891, dtype: int64

In [17]:
# Write your code for dropping the rows here
movies = movies[movies.isnull().sum(axis = 1) <= 5]
movies.shape

(3891, 13)

-  ### Subtask 2.5: Fill NaN values

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [18]:
# Inspect the language column of the dataset
movies['language'].describe()

count        3887
unique         37
top       English
freq         3707
Name: language, dtype: object

In [19]:
# Write your code for filling the NaN values in the 'language' column here
movies['language'].fillna('English', inplace=True)
movies['language'].describe()

count        3891
unique         37
top       English
freq         3711
Name: language, dtype: object

-  ### Subtask 2.6: Check the number of retained rows

You might notice that two of the columns viz. `num_critic_for_reviews` and `actor_1_name` have small percentages of NaN values left. You can let these columns as it is for now. Check the number and percentage of the rows retained after completing all the tasks above.

In [20]:
# Write your code for checking number of retained rows here
len(movies)

3891

In [21]:
100*(len(movies)/5043)

77.15645449137418

**Checkpoint 1:** You might have noticed that we still have around `77%` of the rows!

Checkpoint 1: We have a little over 77% rows i.e., 77.15645449137418

## Task 3: Data Analysis

-  ### Subtask 3.1: Change the unit of columns

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [22]:
# Write your code for unit conversion here
movies['budget_millions']  = movies['budget'].apply(lambda x: x/1000000)
movies['gross_millions']  = movies['gross'].apply(lambda x: x/1000000)
movies.head()

director_name  num_critic_for_reviews        gross  \
0      James Cameron                   723.0  760505847.0   
1     Gore Verbinski                   302.0  309404152.0   
2         Sam Mendes                   602.0  200074175.0   
3  Christopher Nolan                   813.0  448130642.0   
5     Andrew Stanton                   462.0   73058679.0   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
5          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
5                               John Carter            212204   

   num_user_for_reviews language       budget  title_year  imdb_score  \
0                3054.0  English  237000000.0      2009.0         7.9   
1                1238.0  English  300000000.0      2007.0         7.1   
2                 994.0  English  245000000.0      2015.0         6.8   
3                2701.0  English  250000000.0      2012.0         8.5   
5                 738.0  English  263700000.0      2012.0         6.6   

   movie_facebook_likes  budget_millions  gross_millions  
0                 33000            237.0      760.505847  
1                     0            300.0      309.404152  
2                 85000            245.0      200.074175  
3                164000            250.0      448.130642  
5                 24000            263.7       73.058679

-  ### Subtask 3.2: Find the movies with highest profit

    1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
    2. Sort the dataframe using the `profit` column as reference.
    3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [23]:
# Write your code for creating the profit column here
movies['profit'] = movies['gross'] - movies['budget']
movies['profit_millions'] = movies['profit'].apply(lambda x: x/1000000)
movies

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
5        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
5033      Shane Carruth                   143.0     424760.0   
5034   Neill Dela Llana                    35.0      70071.0   
5035   Robert Rodriguez                    56.0    2040920.0   
5037       Edward Burns                    14.0       4584.0   
5042           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
5                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
5033                Drama|Sci-Fi|Thriller    Shane Carruth   
5034                             Thriller      Ian Gamazon   
5035  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
5037                         Comedy|Drama      Kerry Bishé   
5042                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
5                                  John Carter            212204   
...                                         ...              ...   
5033                                    Primer             72639   
5034                                    Cavite               589   
5035                               El Mariachi             52055   
5037                                 Newlyweds              1338   
5042                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
5                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
5033                 371.0  English       7000.0      2004.0         7.0   
5034                  35.0  English       7000.0      2005.0         6.3   
5035                 130.0  Spanish       7000.0      1992.0         6.9   
5037                  14.0  English       9000.0      2011.0         6.4   
5042                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  budget_millions  gross_millions       profit  \
0                    33000         237.0000      760.505847  523505847.0   
1                        0         300.0000      309.404152    9404152.0   
2                    85000         245.0000      200.074175  -44925825.0   
3                   164000         250.0000      448.130642  198130642.0   
5                    24000         263.7000       73.058679 -190641321.0   
...                    ...              ...             ...          ...   
5033                 19000           0.0070        0.424760     417760.0   
5034                    74           0.0070        0.070071      63071.0   
5035                     0           0.0070        2.040920    2033920.0   
5037  

In [24]:
# Write your code for sorting the dataframe here
movies.sort_values(by = 'profit', ascending= False, inplace = True)
movies

director_name  num_critic_for_reviews        gross  \
0        James Cameron                   723.0  760505847.0   
29     Colin Trevorrow                   644.0  652177271.0   
26       James Cameron                   315.0  658672302.0   
3024      George Lucas                   282.0  460935665.0   
3080  Steven Spielberg                   215.0  434949459.0   
...                ...                     ...          ...   
2334   Katsuhiro Ôtomo                   105.0     410388.0   
2323    Hayao Miyazaki                   174.0    2298191.0   
3005      Lajos Koltai                    73.0     195888.0   
3859    Chan-wook Park                   202.0     211667.0   
2988      Joon-ho Bong                   363.0    2201412.0   

                                                 genres         actor_1_name  \
0                       Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29                     Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                                        Drama|Romance    Leonardo DiCaprio   
3024                    Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                                      Family|Sci-Fi         Henry Thomas   
...                                                 ...                  ...   
2334  Action|Adventure|Animation|Family|Sci-Fi|Thriller     William Hootkins   
2323                        Adventure|Animation|Fantasy        Minnie Driver   
3005                                  Drama|Romance|War         Marcell Nagy   
3859                                        Crime|Drama         Min-sik Choi   
2988                         Comedy|Drama|Horror|Sci-Fi            Doona Bae   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
29                        Jurassic World            418214   
26                               Titanic            793059   
3024  Star Wars: Episode IV - A New Hope            911097   
3080          E.T. the Extra-Terrestrial            281842   
...                                   ...              ...   
2334                            Steamboy             13727   
2323                   Princess Mononoke            221552   
3005                            Fateless              5603   
3859                      Lady Vengeance             53508   
2988                            The Host             68883   

      num_user_for_reviews   language        budget  title_year  imdb_score  \
0                   3054.0    English  2.370000e+08      2009.0         7.9   
29                  1290.0    English  1.500000e+08      2015.0         7.0   
26                  2528.0    English  2.000000e+08      1997.0         7.7   
3024                1470.0    English  1.100000e+07      1977.0         8.7   
3080                 515.0    English  1.050000e+07      1982.0         7.9   
...                    ...        ...           ...         ...         ...   
2334                  79.0   Japanese  2.127520e+09      2004.0         6.9   
2323                 570.0   Japanese  2.400000e+09      1997.0         8.4   
3005                  45.0  Hungarian  2.500000e+09      2005.0         7.1   
3859                 131.0     Korean  4.200000e+09      2005.0         7.7   
2988                 279.0     Korean  1.221550e+10      2006.0         7.0   

      movie_facebook_likes  budget_millions  gross_millions        profit  \
0                    33000       237.000000      760.505847  5.235058e+08   
29                  150000       150.000000      652.177271  5.021773e+08   
26                   26000       200.000000      658.672302  4.586723e+08   
3024                 33000        11.000000      460.935665  4.499357e+08   
3080                 34000        10.500000      434.949459  4.244495e+08   
...                    ...              ...             ...           ...   
2334                   973      2127.519898        0.41038

In [25]:
# Write your code to get the top 10 profiting movies here
top10 = movies['movie_title'].head(10)
top10


0                                          Avatar 
29                                 Jurassic World 
26                                        Titanic 
3024           Star Wars: Episode IV - A New Hope 
3080                   E.T. the Extra-Terrestrial 
794                                  The Avengers 
17                                   The Avengers 
509                                 The Lion King 
240     Star Wars: Episode I - The Phantom Menace 
66                                The Dark Knight 
Name: movie_title, dtype: object

-  ### Subtask 3.3: Drop duplicate values

After you found out the top 10 profiting movies, you might have notice a duplicate value. So, it seems like the dataframe has duplicate values as well. Drop the duplicate values from the dataframe and repeat `Subtask 3.2`.

In [26]:
# Write your code for dropping duplicate values here
movies.drop_duplicates(subset = None, keep = 'first', inplace = True)
movies

director_name  num_critic_for_reviews        gross  \
0        James Cameron                   723.0  760505847.0   
29     Colin Trevorrow                   644.0  652177271.0   
26       James Cameron                   315.0  658672302.0   
3024      George Lucas                   282.0  460935665.0   
3080  Steven Spielberg                   215.0  434949459.0   
...                ...                     ...          ...   
2334   Katsuhiro Ôtomo                   105.0     410388.0   
2323    Hayao Miyazaki                   174.0    2298191.0   
3005      Lajos Koltai                    73.0     195888.0   
3859    Chan-wook Park                   202.0     211667.0   
2988      Joon-ho Bong                   363.0    2201412.0   

                                                 genres         actor_1_name  \
0                       Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29                     Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                                        Drama|Romance    Leonardo DiCaprio   
3024                    Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                                      Family|Sci-Fi         Henry Thomas   
...                                                 ...                  ...   
2334  Action|Adventure|Animation|Family|Sci-Fi|Thriller     William Hootkins   
2323                        Adventure|Animation|Fantasy        Minnie Driver   
3005                                  Drama|Romance|War         Marcell Nagy   
3859                                        Crime|Drama         Min-sik Choi   
2988                         Comedy|Drama|Horror|Sci-Fi            Doona Bae   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
29                        Jurassic World            418214   
26                               Titanic            793059   
3024  Star Wars: Episode IV - A New Hope            911097   
3080          E.T. the Extra-Terrestrial            281842   
...                                   ...              ...   
2334                            Steamboy             13727   
2323                   Princess Mononoke            221552   
3005                            Fateless              5603   
3859                      Lady Vengeance             53508   
2988                            The Host             68883   

      num_user_for_reviews   language        budget  title_year  imdb_score  \
0                   3054.0    English  2.370000e+08      2009.0         7.9   
29                  1290.0    English  1.500000e+08      2015.0         7.0   
26                  2528.0    English  2.000000e+08      1997.0         7.7   
3024                1470.0    English  1.100000e+07      1977.0         8.7   
3080                 515.0    English  1.050000e+07      1982.0         7.9   
...                    ...        ...           ...         ...         ...   
2334                  79.0   Japanese  2.127520e+09      2004.0         6.9   
2323                 570.0   Japanese  2.400000e+09      1997.0         8.4   
3005                  45.0  Hungarian  2.500000e+09      2005.0         7.1   
3859                 131.0     Korean  4.200000e+09      2005.0         7.7   
2988                 279.0     Korean  1.221550e+10      2006.0         7.0   

      movie_facebook_likes  budget_millions  gross_millions        profit  \
0                    33000       237.000000      760.505847  5.235058e+08   
29                  150000       150.000000      652.177271  5.021773e+08   
26                   26000       200.000000      658.672302  4.586723e+08   
3024                 33000        11.000000      460.935665  4.499357e+08   
3080                 34000        10.500000      434.949459  4.244495e+08   
...                    ...              ...             ...           ...   
2334                   973      2127.519898        0.41038

In [27]:
# Write code for repeating subtask 2 here
top10 = movies['movie_title'].head(10)
top10

0                                          Avatar 
29                                 Jurassic World 
26                                        Titanic 
3024           Star Wars: Episode IV - A New Hope 
3080                   E.T. the Extra-Terrestrial 
794                                  The Avengers 
509                                 The Lion King 
240     Star Wars: Episode I - The Phantom Menace 
66                                The Dark Knight 
439                              The Hunger Games 
Name: movie_title, dtype: object

**Checkpoint 2:** You might spot two movies directed by `James Cameron` in the list.

Checkpoint 2: Two movies directed by James Cameron in the list are Avatar and Titanic

-  ### Subtask 3.4: Find IMDb Top 250

    1. Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000.
Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.
    2. Extract all the movies in the `IMDb_Top_250` dataframe which are not in the English language and store them in a new dataframe named `Top_Foreign_Lang_Film`.

In [28]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'
IMDb_Top_250 = movies.sort_values(by = 'imdb_score', ascending = False)
IMDb_Top_250 = IMDb_Top_250.loc[IMDb_Top_250.num_voted_users > 25000]
IMDb_Top_250 = IMDb_Top_250.iloc[:250, ]
IMDb_Top_250['Rank'] = range(1,251)
IMDb_Top_250

director_name  num_critic_for_reviews        gross  \
1937        Frank Darabont                   199.0   28341469.0   
3466  Francis Ford Coppola                   208.0  134821952.0   
66       Christopher Nolan                   645.0  533316061.0   
2837  Francis Ford Coppola                   149.0   57300000.0   
339          Peter Jackson                   328.0  377019252.0   
...                    ...                     ...          ...   
788          Cameron Crowe                   149.0   32522352.0   
99           Peter Jackson                   645.0  303001229.0   
1606       Nick Cassavetes                   177.0      64286.0   
1735         James Mangold                   291.0  119518352.0   
639           Michael Mann                   209.0   28965197.0   

                              genres            actor_1_name  \
1937                     Crime|Drama          Morgan Freeman   
3466                     Crime|Drama               Al Pacino   
66       Action|Crime|Drama|Thriller          Christian Bale   
2837                     Crime|Drama          Robert De Niro   
339   Action|Adventure|Drama|Fantasy           Orlando Bloom   
...                              ...                     ...   
788     Adventure|Comedy|Drama|Music  Philip Seymour Hoffman   
99                 Adventure|Fantasy            Aidan Turner   
1606                   Drama|Romance            Ryan Gosling   
1735   Biography|Drama|Music|Romance   Sandra Ellis Lafferty   
639         Biography|Drama|Thriller               Al Pacino   

                                         movie_title  num_voted_users  \
1937                       The Shawshank Redemption           1689764   
3466                                  The Godfather           1155770   
66                                  The Dark Knight           1676169   
2837                         The Godfather: Part II            790926   
339   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
788                                   Almost Famous            207287   
99                The Hobbit: An Unexpected Journey            637246   
1606                                   The Notebook            396396   
1735                                  Walk the Line            188637   
639                                     The Insider            133526   

      num_user_for_reviews language       budget  title_year  imdb_score  \
1937                4144.0  English   25000000.0      1994.0         9.3   
3466                2238.0  English    6000000.0      1972.0         9.2   
66                  4667.0  English  185000000.0      2008.0         9.0   
2837                 650.0  English   13000000.0      1974.0         9.0   
339                 3189.0  English   94000000.0      2003.0         8.9   
...                    ...      ...          ...         ...         ...   
788                  822.0  English   60000000.0      2000.0         7.9   
99                  1367.0  English  180000000.0      2012.0         7.9   
1606                1111.0  English   29000000.0      2004.0         7.9   
1735                 815.0  English   28000000.0      2005.0         7.9   
639                  521.0  English   68000000.0      1999.0         7.9   

      movie_facebook_likes  budget_millions  gross_millions       profit  \
1937                108000             25.0       28.341469    3341469.0   
3466                 43000              6.0      134.821952  128821952.0   
66                   37000            185.0      533.316061  348316061.0   
2837                 14000             13.0       57.300000   44300000.0   
339                  16000             94.0      377.019252  283019252.0   
...                    ...              ...             ...          ...   
788                  15000             60.0       32.522352  -27477648.0   
99                  166000            180

In [29]:
Top_Foreign_Lang_Film = IMDb_Top_250.loc[IMDb_Top_250['language'] != 'English']
Top_Foreign_Lang_Film

director_name  num_critic_for_reviews       gross  \
4498                      Sergio Leone                   181.0   6100000.0   
4029                Fernando Meirelles                   214.0   7563397.0   
4747                    Akira Kurosawa                   153.0    269061.0   
2373                    Hayao Miyazaki                   246.0  10049886.0   
4259  Florian Henckel von Donnersmarck                   215.0  11284657.0   
4921                      Majid Majidi                    46.0    925402.0   
4105                    Chan-wook Park                   305.0   2181290.0   
1298                Jean-Pierre Jeunet                   242.0  33201661.0   
1329                    S.S. Rajamouli                    44.0   6498000.0   
2323                    Hayao Miyazaki                   174.0   2298191.0   
2970                 Wolfgang Petersen                    96.0  11433134.0   
4659                    Asghar Farhadi                   354.0   7098492.0   
4033                 Thomas Vinterberg                   349.0    610968.0   
2829               Oliver Hirschbiegel                   192.0   5501940.0   
2734                        Fritz Lang                   260.0     26435.0   
3550                  Denis Villeneuve                   226.0   6857096.0   
4000              Juan José Campanella                   262.0  20167424.0   
2047                    Hayao Miyazaki                   212.0   4710455.0   
2551                Guillermo del Toro                   406.0  37623143.0   
3553                      José Padilha                   142.0      8060.0   
2914                       Je-kyu Kang                    86.0   1110186.0   
2830                Alejandro Amenábar                   157.0   2086345.0   
4267             Alejandro G. Iñárritu                   157.0   5383834.0   
3423                   Katsuhiro Ôtomo                   150.0    439162.0   
4461                 Thomas Vinterberg                    98.0   1647780.0   
3344                       Karan Johar                   210.0   4018695.0   
4284                        Ari Folman                   231.0   2283276.0   
3456                 Vincent Paronnaud                   242.0   4443403.0   
4144                     Walter Salles                    71.0   5595428.0   
4897                      Sergio Leone                   122.0   3500000.0   
3677              Christophe Barratier                   112.0   3629758.0   
4640                   Cristian Mungiu                   233.0   1185783.0   
4415                  Fabián Bielinsky                    94.0   1221261.0   
2863                    Clint Eastwood                   251.0  13753931.0   
3510                       Yash Chopra                    29.0   2921738.0   
3264                    Michael Haneke                   447.0    225377.0   

                                                 genres  \
4498                                            Western   
4029                                        Crime|Drama   
4747                             Action|Adventure|Drama   
2373                 Adventure|Animation|Family|Fantasy   
4259                                     Drama|Thriller   
4921                                       Drama|Family   
4105                             Drama|Mystery|Thriller   
1298                                     Comedy|Romance   
1329                 Action|Adventure|Drama|Fantasy|War   
2323                        Adventure|Animation|Fantasy   
2970                       Adventure|Drama|Thriller|War   
4659                                      Drama|Mystery   
4033                                              Drama   
2829                        Biography|Drama|History|War   
2734                                       Drama|Sci-Fi   
3550                                  Drama|Mystery|War   
4000                             Drama|Mystery|Thriller   
2047                 Adventure|Animation|Family|Fantasy   
2551             

**Checkpoint 3:** Can you spot `Veer-Zaara` in the dataframe?


Checkpoint 3: Veer Zaara is on rank 228 in the top 250 list.

- ### Subtask 3.5: Find the best directors

    1. Group the dataframe using the `director_name` column.
    2. Find out the top 10 directors for whom the mean of `imdb_score` is the highest and store them in a new dataframe `top10director`. 

In [31]:
# Write your code for extracting the top 10 directors here
top10director = movies.groupby('director_name').imdb_score.mean().reset_index()
top10director.sort_values(by = 'imdb_score', ascending = False, inplace = True)
top10director = top10director.head(10)
top10director

director_name  imdb_score
216         Charles Chaplin    8.600000
1675              Tony Kaye    8.600000
45         Alfred Hitchcock    8.500000
1440             Ron Fricke    8.500000
302         Damien Chazelle    8.500000
1017           Majid Majidi    8.500000
1498           Sergio Leone    8.433333
260       Christopher Nolan    8.425000
1467         S.S. Rajamouli    8.400000
1035  Marius A. Markevicius    8.400000

**Checkpoint 4:** No surprises that `Damien Chazelle` (director of Whiplash and La La Land) is in this list.

-  ### Subtask 3.6: Find popular genres

You might have noticed the `genres` column in the dataframe with all the genres of the movies seperated by a pipe (`|`). Out of all the movie genres, the first two are most significant for any film.

1. Extract the first two genres from the `genres` column and store them in two new columns: `genre_1` and `genre_2`. Some of the movies might have only one genre. In such cases, extract the single genre into both the columns, i.e. for such movies the `genre_2` will be the same as `genre_1`.
2. Group the dataframe using `genre_1` as the primary column and `genre_2` as the secondary column.
3. Find out the 5 most popular combo of genres by finding the mean of the gross values using the `gross` column and store them in a new dataframe named `PopGenre`.

In [32]:
# Write your code for extracting the first two genres of each movie here
genres = movies['genres'].str.split('|')
movies['genre_1'] = genres.apply(lambda x: x[0])
movies['genre_2'] = genres.apply(lambda x:  x[1] if len(x) > 1 else x[0])
movies

director_name  num_critic_for_reviews        gross  \
0        James Cameron                   723.0  760505847.0   
29     Colin Trevorrow                   644.0  652177271.0   
26       James Cameron                   315.0  658672302.0   
3024      George Lucas                   282.0  460935665.0   
3080  Steven Spielberg                   215.0  434949459.0   
...                ...                     ...          ...   
2334   Katsuhiro Ôtomo                   105.0     410388.0   
2323    Hayao Miyazaki                   174.0    2298191.0   
3005      Lajos Koltai                    73.0     195888.0   
3859    Chan-wook Park                   202.0     211667.0   
2988      Joon-ho Bong                   363.0    2201412.0   

                                                 genres         actor_1_name  \
0                       Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29                     Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                                        Drama|Romance    Leonardo DiCaprio   
3024                    Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                                      Family|Sci-Fi         Henry Thomas   
...                                                 ...                  ...   
2334  Action|Adventure|Animation|Family|Sci-Fi|Thriller     William Hootkins   
2323                        Adventure|Animation|Fantasy        Minnie Driver   
3005                                  Drama|Romance|War         Marcell Nagy   
3859                                        Crime|Drama         Min-sik Choi   
2988                         Comedy|Drama|Horror|Sci-Fi            Doona Bae   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
29                        Jurassic World            418214   
26                               Titanic            793059   
3024  Star Wars: Episode IV - A New Hope            911097   
3080          E.T. the Extra-Terrestrial            281842   
...                                   ...              ...   
2334                            Steamboy             13727   
2323                   Princess Mononoke            221552   
3005                            Fateless              5603   
3859                      Lady Vengeance             53508   
2988                            The Host             68883   

      num_user_for_reviews   language        budget  title_year  imdb_score  \
0                   3054.0    English  2.370000e+08      2009.0         7.9   
29                  1290.0    English  1.500000e+08      2015.0         7.0   
26                  2528.0    English  2.000000e+08      1997.0         7.7   
3024                1470.0    English  1.100000e+07      1977.0         8.7   
3080                 515.0    English  1.050000e+07      1982.0         7.9   
...                    ...        ...           ...         ...         ...   
2334                  79.0   Japanese  2.127520e+09      2004.0         6.9   
2323                 570.0   Japanese  2.400000e+09      1997.0         8.4   
3005                  45.0  Hungarian  2.500000e+09      2005.0         7.1   
3859                 131.0     Korean  4.200000e+09      2005.0         7.7   
2988                 279.0     Korean  1.221550e+10      2006.0         7.0   

      movie_facebook_likes  budget_millions  gross_millions        profit  \
0                    33000       237.000000      760.505847  5.235058e+08   
29                  150000       150.000000      652.177271  5.021773e+08   
26                   26000       200.000000      658.672302  4.586723e+08   
3024                 33000        11.000000      460.935665  4.499357e+08   
3080                 34000        10.500000      434.949459  4.244495e+08   
...                    ...              ...             ...           ...   
2334                   973      2127.519898        0.41038

In [33]:
movies_by_segment = movies.groupby(['genre_1', 'genre_2'])


In [34]:
PopGenre = pd.DataFrame(movies_by_segment['gross'].mean()).sort_values(by = 'gross', ascending = False)
PopGenre['gross_millions'] = PopGenre['gross'].apply(lambda x: x/1000000)
PopGenre.drop(columns = ['gross'], inplace=True)
PopGenre.head(5)


gross_millions
genre_1   genre_2                  
Family    Sci-Fi         434.949459
Adventure Sci-Fi         228.627758
          Family         118.919540
          Animation      116.998550
Action    Adventure      109.595465

**Checkpoint 5:** Well, as it turns out. `Family + Sci-Fi` is the most popular combo of genres out there!

-  ### Subtask 3.7: Find the critic-favorite and audience-favorite actors

    1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
    2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
    3. Group the combined dataframe using the `actor_1_name` column.
    4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [35]:
# Write your code for creating three new dataframes here
Meryl_Streep = movies[movies['actor_1_name'] == 'Meryl Streep']
Meryl_Streep

director_name  num_critic_for_reviews        gross  \
1408   David Frankel                   208.0  124732962.0   
1575  Sydney Pollack                    66.0   87100000.0   
1204     Nora Ephron                   252.0   94125426.0   
1618   David Frankel                   234.0   63536011.0   
410     Nancy Meyers                   187.0  112703470.0   
2781  Phyllida Lloyd                   331.0   29959436.0   
1925  Stephen Daldry                   174.0   41597830.0   
3135   Robert Altman                   211.0   20338609.0   
1106   Curtis Hanson                    42.0   46815748.0   
1674   Carl Franklin                    64.0   23209440.0   
1483  Robert Redford                   227.0   14998070.0   

                               genres  actor_1_name  \
1408             Comedy|Drama|Romance  Meryl Streep   
1575          Biography|Drama|Romance  Meryl Streep   
1204          Biography|Drama|Romance  Meryl Streep   
1618             Comedy|Drama|Romance  Meryl Streep   
410              Comedy|Drama|Romance  Meryl Streep   
2781          Biography|Drama|History  Meryl Streep   
1925                    Drama|Romance  Meryl Streep   
3135               Comedy|Drama|Music  Meryl Streep   
1106  Action|Adventure|Crime|Thriller  Meryl Streep   
1674                            Drama  Meryl Streep   
1483               Drama|Thriller|War  Meryl Streep   

                    movie_title  num_voted_users  num_user_for_reviews  \
1408     The Devil Wears Prada            286178                 631.0   
1575             Out of Africa             52339                 200.0   
1204             Julie & Julia             79264                 277.0   
1618              Hope Springs             34258                 178.0   
410           It's Complicated             69860                 214.0   
2781             The Iron Lady             82327                 350.0   
1925                 The Hours            102123                 660.0   
3135  A Prairie Home Companion             19655                 280.0   
1106            The River Wild             32544                  69.0   
1674            One True Thing              9283                 112.0   
1483           Lions for Lambs             41170                 298.0   

     language      budget  title_year  imdb_score  movie_facebook_likes  \
1408  English  35000000.0      2006.0         6.8                     0   
1575  English  31000000.0      1985.0         7.2                     0   
1204  English  40000000.0      2009.0         7.0                 13000   
1618  English  30000000.0      2012.0         6.3                     0   
410   English  85000000.0      2009.0         6.6                     0   
2781  English  13000000.0      2011.0         6.4                 18000   
1925  English  25000000.0      2002.0         7.6                     0   
3135  English  10000000.0      2006.0         6.8                   683   
1106  English  45000000.0      1994.0         6.3                     0   
1674  English  30000000.0      1998.0         7.0                   592   
1483  English  35000000.0      2007.0         6.2                     0   

      budget_millions  gross_millions      profit  profit_millions    genre_1  \
1408             35.0      124.732962  89732962.0        89.732962     Comedy   
1575             31.0       87.100000  56100000.0        56.100000  Biography   
1204             40.0       94.125426  54125426.0        54.125426  Biography   
1618             30.0       63.536011  33536011.0        33.536011     Comedy   
410              85.0      112.703470  27703470.0        27.703470     Comedy   
2781             13.0       29.959436  16959436.0        16.959436  Biography   
1925             25.0       41.597830  16597830.0        16.597830      Drama   
3135             10.0       20.338609  10338609.0        10.338609     Comedy   
1106             45.0       46.815748   1815748.0         1.815748     Action   
1674         

In [36]:
Leo_Caprio = movies[movies['actor_1_name'] == 'Leonardo DiCaprio']
Leo_Caprio

director_name  num_critic_for_reviews        gross  \
26            James Cameron                   315.0  658672302.0   
97        Christopher Nolan                   642.0  292568851.0   
911        Steven Spielberg                   194.0  164435221.0   
296       Quentin Tarantino                   765.0  162804648.0   
179   Alejandro G. Iñárritu                   556.0  183635922.0   
452         Martin Scorsese                   490.0  127968405.0   
361         Martin Scorsese                   352.0  132373442.0   
50             Baz Luhrmann                   490.0  144812796.0   
3476           Baz Luhrmann                   490.0  144812796.0   
2757           Baz Luhrmann                   106.0   46338728.0   
1422        Randall Wallace                    83.0   56876365.0   
308         Martin Scorsese                   606.0  116866727.0   
1453         Clint Eastwood                   392.0   37304950.0   
257         Martin Scorsese                   267.0  102608827.0   
2067             Jerry Zaks                    45.0   12782508.0   
990             Danny Boyle                   118.0   39778599.0   
1114             Sam Mendes                   323.0   22877808.0   
1560              Sam Raimi                    63.0   18636537.0   
326         Martin Scorsese                   233.0   77679638.0   
641            Ridley Scott                   238.0   39380442.0   
307            Edward Zwick                   166.0   57366262.0   

                                genres       actor_1_name  \
26                       Drama|Romance  Leonardo DiCaprio   
97    Action|Adventure|Sci-Fi|Thriller  Leonardo DiCaprio   
911              Biography|Crime|Drama  Leonardo DiCaprio   
296                      Drama|Western  Leonardo DiCaprio   
179   Adventure|Drama|Thriller|Western  Leonardo DiCaprio   
452                   Mystery|Thriller  Leonardo DiCaprio   
361               Crime|Drama|Thriller  Leonardo DiCaprio   
50                       Drama|Romance  Leonardo DiCaprio   
3476                     Drama|Romance  Leonardo DiCaprio   
2757                     Drama|Romance  Leonardo DiCaprio   
1422                  Action|Adventure  Leonardo DiCaprio   
308       Biography|Comedy|Crime|Drama  Leonardo DiCaprio   
1453             Biography|Crime|Drama  Leonardo DiCaprio   
257                    Biography|Drama  Leonardo DiCaprio   
2067                             Drama  Leonardo DiCaprio   
990           Adventure|Drama|Thriller  Leonardo DiCaprio   
1114                     Drama|Romance  Leonardo DiCaprio   
1560           Action|Thriller|Western  Leonardo DiCaprio   
326                        Crime|Drama  Leonardo DiCaprio   
641              Action|Drama|Thriller  Leonardo DiCaprio   
307           Adventure|Drama|Thriller  Leonardo DiCaprio   

                    movie_title  num_voted_users  num_user_for_reviews  \
26                     Titanic            793059                2528.0   
97                   Inception           1468200                2803.0   
911        Catch Me If You Can            525801                 667.0   
296           Django Unchained            955174                1193.0   
179               The Revenant            406020                1188.0   
452             Shutter Island            786092                 964.0   
361               The Departed            873649                2054.0   
50            The Great Gatsby            362912                 753.0   
3476          The Great Gatsby            362933                 753.0   
2757            Romeo + Juliet            167750                 506.0   
1422  The Man in the Iron Mask            125219                 244.0   
308    The Wolf of Wall Street            780588                1138.0   
1453                  J. Edgar            102728                 279.0   
257                The Aviator            264318                 799.0   
2067             Marvin's Room             20163                

In [37]:
Brad_Pitt = movies[movies['actor_1_name'] == 'Brad Pitt']
Brad_Pitt

director_name  num_critic_for_reviews        gross  \
400       Steven Soderbergh                   186.0  183405771.0   
255              Doug Liman                   233.0  186336103.0   
940             Neil Jordan                   120.0  105264608.0   
470              David Ayer                   406.0   85707116.0   
254       Steven Soderbergh                   198.0  125531634.0   
2204  Alejandro G. Iñárritu                   285.0   34300771.0   
2682         Andrew Dominik                   414.0   14938570.0   
2898             Tony Scott                   122.0   12281500.0   
2333    Angelina Jolie Pitt                   131.0     531009.0   
1490        Terrence Malick                   584.0   13303319.0   
101           David Fincher                   362.0  127490802.0   
683           David Fincher                   315.0   37023395.0   
1722         Andrew Dominik                   273.0    3904982.0   
611     Jean-Jacques Annaud                    76.0   37901509.0   
792         Patrick Gilmore                    98.0   26288320.0   
147       Wolfgang Petersen                   220.0  133228348.0   
382              Tony Scott                   142.0      26871.0   

                                                 genres actor_1_name  \
400                                      Crime|Thriller    Brad Pitt   
255                Action|Comedy|Crime|Romance|Thriller    Brad Pitt   
940                                Drama|Fantasy|Horror    Brad Pitt   
470                                    Action|Drama|War    Brad Pitt   
254                                      Crime|Thriller    Brad Pitt   
2204                                              Drama    Brad Pitt   
2682                                     Crime|Thriller    Brad Pitt   
2898                Action|Crime|Drama|Romance|Thriller    Brad Pitt   
2333                                      Drama|Romance    Brad Pitt   
1490                                      Drama|Fantasy    Brad Pitt   
101                               Drama|Fantasy|Romance    Brad Pitt   
683                                               Drama    Brad Pitt   
1722              Biography|Crime|Drama|History|Western    Brad Pitt   
611               Adventure|Biography|Drama|History|War    Brad Pitt   
792   Adventure|Animation|Comedy|Drama|Family|Fantas...    Brad Pitt   
147                                           Adventure    Brad Pitt   
382                               Action|Crime|Thriller    Brad Pitt   

                                            movie_title  num_voted_users  \
400                                     Ocean's Eleven            402645   
255                                   Mr. & Mrs. Smith            348861   
940   Interview with the Vampire: The Vampire Chroni...           239752   
470                                               Fury            303185   
254                                     Ocean's Twelve            284852   
2204                                             Babel            243799   
2682                               Killing Them Softly            111625   
2898                                      True Romance            163492   
2333                                        By the Sea              7976   
1490                                  The Tree of Life            136367   
101                The Curious Case of Benjamin Button            459346   
683                                         Fight Club           1347461   
1722  The Assassination of Jesse James by the Coward...           136104   
611                               Seven Years in Tibet             96385   
792                   Sinbad: Legend of the Seven Seas             36144   
147                                               Troy            381672   
382                                           Spy Game            121259   

      num_user_for_reviews language       budget  title_year  imdb_score  \
400                  845.0  English   85000000.

In [38]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt])
Combined

director_name  num_critic_for_reviews        gross  \
1408          David Frankel                   208.0  124732962.0   
1575         Sydney Pollack                    66.0   87100000.0   
1204            Nora Ephron                   252.0   94125426.0   
1618          David Frankel                   234.0   63536011.0   
410            Nancy Meyers                   187.0  112703470.0   
2781         Phyllida Lloyd                   331.0   29959436.0   
1925         Stephen Daldry                   174.0   41597830.0   
3135          Robert Altman                   211.0   20338609.0   
1106          Curtis Hanson                    42.0   46815748.0   
1674          Carl Franklin                    64.0   23209440.0   
1483         Robert Redford                   227.0   14998070.0   
26            James Cameron                   315.0  658672302.0   
97        Christopher Nolan                   642.0  292568851.0   
911        Steven Spielberg                   194.0  164435221.0   
296       Quentin Tarantino                   765.0  162804648.0   
179   Alejandro G. Iñárritu                   556.0  183635922.0   
452         Martin Scorsese                   490.0  127968405.0   
361         Martin Scorsese                   352.0  132373442.0   
50             Baz Luhrmann                   490.0  144812796.0   
3476           Baz Luhrmann                   490.0  144812796.0   
2757           Baz Luhrmann                   106.0   46338728.0   
1422        Randall Wallace                    83.0   56876365.0   
308         Martin Scorsese                   606.0  116866727.0   
1453         Clint Eastwood                   392.0   37304950.0   
257         Martin Scorsese                   267.0  102608827.0   
2067             Jerry Zaks                    45.0   12782508.0   
990             Danny Boyle                   118.0   39778599.0   
1114             Sam Mendes                   323.0   22877808.0   
1560              Sam Raimi                    63.0   18636537.0   
326         Martin Scorsese                   233.0   77679638.0   
641            Ridley Scott                   238.0   39380442.0   
307            Edward Zwick                   166.0   57366262.0   
400       Steven Soderbergh                   186.0  183405771.0   
255              Doug Liman                   233.0  186336103.0   
940             Neil Jordan                   120.0  105264608.0   
470              David Ayer                   406.0   85707116.0   
254       Steven Soderbergh                   198.0  125531634.0   
2204  Alejandro G. Iñárritu                   285.0   34300771.0   
2682         Andrew Dominik                   414.0   14938570.0   
2898             Tony Scott                   122.0   12281500.0   
2333    Angelina Jolie Pitt                   131.0     531009.0   
1490        Terrence Malick                   584.0   13303319.0   
101           David Fincher                   362.0  127490802.0   
683           David Fincher                   315.0   37023395.0   
1722         Andrew Dominik                   273.0    3904982.0   
611     Jean-Jacques Annaud                    76.0   37901509.0   
792         Patrick Gilmore                    98.0   26288320.0   
147       Wolfgang Petersen                   220.0  133228348.0   
382              Tony Scott                   142.0      26871.0   

                                                 genres       actor_1_name  \
1408                               Comedy|Drama|Romance       Meryl Streep   
1575                            Biography|Drama|Romance       Meryl Streep   
1204                            Biography|Drama|Romance       Meryl Streep   
1618                               Comedy|Drama|Romance       Meryl Streep   
410                                Comedy|Drama|Romance       Meryl Streep   
2781                            Biography|Drama|History       Meryl Streep   
1925                                      Drama|Romance       Meryl

In [39]:
# Check for Null Values in the Combined Dataframe
Combined.isnull().sum()

director_name             0
num_critic_for_reviews    0
gross                     0
genres                    0
actor_1_name              0
movie_title               0
num_voted_users           0
num_user_for_reviews      0
language                  0
budget                    0
title_year                0
imdb_score                0
movie_facebook_likes      0
budget_millions           0
gross_millions            0
profit                    0
profit_millions           0
genre_1                   0
genre_2                   0
dtype: int64

In [40]:
# Write your code for grouping the combined dataframe here
Combined_by_Actors = Combined.groupby('actor_1_name')
Combined_by_Actors


In [42]:
# Write the code for finding the mean of critic reviews and audience reviews here
Actor_Reviews = Combined.groupby('actor_1_name').agg({'num_critic_for_reviews':'mean', 'num_user_for_reviews':'mean'})
Actor_Reviews

num_critic_for_reviews  num_user_for_reviews
actor_1_name                                                   
Brad Pitt                      245.000000            742.352941
Leonardo DiCaprio              330.190476            914.476190
Meryl Streep                   181.454545            297.181818

**Checkpoint 6:** `Leonardo` has aced both the lists!